File: app/src/main/java/ua/com/programmer/simpleremote/ScannerActivity.java
package ua.com.programmer.simpleremote;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.Preview;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.content.ContextCompat;

import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.google.common.util.concurrent.ListenableFuture;

import java.util.ArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import ua.com.programmer.simpleremote.specialItems.Cache;
import ua.com.programmer.simpleremote.specialItems.DataBaseItem;
import ua.com.programmer.simpleremote.utility.BarcodeFoundListener;
import ua.com.programmer.simpleremote.utility.BarcodeImageAnalyzer;
import ua.com.programmer.simpleremote.utility.Utils;

public class ScannerActivity extends AppCompatActivity implements DataLoader.DataLoaderListener{

    private PreviewView cameraView;
    private ListenableFuture<ProcessCameraProvider> cameraProvider;
    private ExecutorService cameraExecutor;

    private TextView textValue;
    private TextView textDescription;
    private ProgressBar progressBar;
    private DataBaseItem dataBaseItem;
    private String documentGUID;

    private final Utils utils = new Utils();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_scanner);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null){
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        setTitle(R.string.header_scanner);

        Intent intent = getIntent();
        documentGUID = intent.getStringExtra("document");

        textValue = findViewById(R.id.text_value);
        textDescription = findViewById(R.id.text_item_description);
        textDescription.setText("");
        progressBar = findViewById(R.id.progress_bar);

        cameraExecutor = Executors.newSingleThreadExecutor();
        cameraView = findViewById(R.id.camera_view);
        cameraProvider = ProcessCameraProvider.getInstance(this);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(android.Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                requestPermissions(new String[]{Manifest.permission.CAMERA}, 1);
            }else {
                setupCamera();
            }
        }else {
            setupCamera();
        }

        TextView buttonRepeat = findViewById(R.id.button_repeat);
        buttonRepeat.setOnClickListener((View v) -> startCamera());

        TextView buttonConfirm = findViewById(R.id.button_confirm);
        buttonConfirm.setOnClickListener((View v) -> confirmAddItem());

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length>0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            finish();
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    private void setupCamera(){
        if (cameraProvider == null) {
            //textView.setText(R.string.error_detector);
            return;
        }

        BarcodeImageAnalyzer barcodeImageAnalyzer = new BarcodeImageAnalyzer(new BarcodeFoundListener() {
            @Override
            public void onBarcodeFound(String barCode, int format) {
                onBarcodeReceived(barCode);
            }

            @Override
            public void onCodeNotFound(String error) {
                utils.debug("on code not found: "+error);
            }
        });

        cameraProvider.addListener(() -> {
            try {
                ProcessCameraProvider provider = cameraProvider.get();
                Preview preview = new Preview.Builder()
                        .build();
                preview.setSurfaceProvider(cameraView.getSurfaceProvider());

                ImageAnalysis imageAnalysis = new ImageAnalysis.Builder()
                        .build();
                imageAnalysis.setAnalyzer(cameraExecutor, barcodeImageAnalyzer);

                CameraSelector cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA;

                try{
                    provider.unbindAll();
                    provider.bindToLifecycle(this, cameraSelector, imageAnalysis, preview);
                }catch (Exception e){
                    utils.debug("bind camera provider error; "+e.getMessage());
                }

            } catch (Exception e) {
                utils.debug("Error starting camera " + e.getMessage());
            }
        }, ContextCompat.getMainExecutor(this));

    }

    private void startCamera(){
        dataBaseItem = null;
        textValue.setText("");
        textDescription.setText("");
        cameraProvider = null;
        try {
            textValue.setText(R.string.scanning);
            setupCamera();
        }catch (SecurityException se) {
            textValue.setText(R.string.error_no_permission);
        }catch (Exception e){
            textValue.setText(e.toString());
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) onBackPressed();
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDataLoaded(ArrayList<DataBaseItem> dataItems) {
        progressBar.setVisibility(View.GONE);
        if (dataItems.size() > 0) {
            dataBaseItem = dataItems.get(0);
            textDescription.setText(dataBaseItem.getString("description"));
            textDescription.setTextColor(ContextCompat.getColor(this,R.color.colorPrimaryDark));
            confirmAddItem();
        }else {
            dataBaseItem = null;
            textDescription.setText(R.string.warn_no_barcode);
            textDescription.setTextColor(ContextCompat.getColor(this,R.color.colorAccent));
        }

    }

    @Override
    public void onDataLoaderError(String error) {
        progressBar.setVisibility(View.GONE);
        Toast.makeText(this, R.string.toast_error, Toast.LENGTH_SHORT).show();
    }

    private void onBarcodeReceived(String barcodeValue){
        textValue.setText(barcodeValue);
        progressBar.setVisibility(View.VISIBLE);

        DataBaseItem barcodeParameters = new DataBaseItem();
        barcodeParameters.put("value",barcodeValue);
        barcodeParameters.put("guid",documentGUID);

        DataLoader dataLoader = new DataLoader(this);
        dataLoader.getItemWithBarcode(barcodeParameters);
    }

    private void confirmAddItem(){
        Intent intent = getIntent();
        if (dataBaseItem != null) {
            intent.putExtra("cacheKey", Cache.getInstance().put(dataBaseItem));
        }else{
            intent.putExtra("cacheKey", "");
        }
        setResult(RESULT_OK, intent);
        finish();
    }
}

File: app/src/main/java/ua/com/programmer/simpleremote/deprecated/ScannerActivity.kt
package ua.com.programmer.simpleremote.deprecated

import android.Manifest
import android.content.pm.PackageManager
import android.media.AudioManager
import android.media.ToneGenerator
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import com.google.common.util.concurrent.ListenableFuture
import ua.com.programmer.simpleremote.R
import ua.com.programmer.simpleremote.repository.DataLoader
import ua.com.programmer.simpleremote.deprecated.specialItems.Cache
import ua.com.programmer.simpleremote.deprecated.specialItems.DataBaseItem
import ua.com.programmer.simpleremote.deprecated.utility.BarcodeFoundListener
import ua.com.programmer.simpleremote.deprecated.utility.BarcodeImageAnalyzer
import ua.com.programmer.simpleremote.deprecated.utility.Utils
import java.lang.Exception
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class ScannerActivity : AppCompatActivity(), DataLoader.DataLoaderListener {
    private var cameraView: PreviewView? = null
    private var cameraProvider: ListenableFuture<ProcessCameraProvider>? = null
    private var cameraExecutor: ExecutorService? = null

    private var textValue: TextView? = null
    private var textDescription: TextView? = null
    private var progressBar: ProgressBar? = null
    private var dataBaseItem: DataBaseItem? = null
    private var documentGUID: String? = null

    private val utils = Utils()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_scanner)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        setTitle(R.string.header_scanner)

        val intent = getIntent()
        documentGUID = intent.getStringExtra("document")

        textValue = findViewById<TextView>(R.id.text_value)
        textDescription = findViewById<TextView>(R.id.text_item_description)
        textDescription!!.text = ""
        progressBar = findViewById<ProgressBar>(R.id.progress_bar)

        cameraExecutor = Executors.newSingleThreadExecutor()
        cameraView = findViewById<PreviewView>(R.id.camera_view)
        try {
            cameraProvider = ProcessCameraProvider.Companion.getInstance(this)
        } catch (e: Exception) {
            utils.debug("ScannerActivity: create camera provider: $e")
        }

        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf<String>(Manifest.permission.CAMERA), 1)
        } else {
            setupCamera()
        }

        val buttonRepeat = findViewById<TextView>(R.id.button_repeat)
        buttonRepeat.setOnClickListener(View.OnClickListener { v: View? -> startCamera() })

        val buttonConfirm = findViewById<TextView>(R.id.button_confirm)
        buttonConfirm.setOnClickListener(View.OnClickListener { v: View? -> confirmAddItem() })
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String?>,
        grantResults: IntArray
    ) {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            finish()
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    private fun setupCamera() {
        if (cameraProvider == null) {
            Toast.makeText(this, R.string.toast_error, Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        val barcodeImageAnalyzer = BarcodeImageAnalyzer(object : BarcodeFoundListener {
            override fun onBarcodeFound(barCode: String?, format: Int) {
                onBarcodeReceived(barCode)
            }

            override fun onCodeNotFound(error: String?) {
                utils.debug("BarcodeImageAnalyzer onCodeNotFound; $error")
            }
        })

        cameraProvider!!.addListener(Runnable {
            try {
                val provider = cameraProvider!!.get()
                val preview = Preview.Builder()
                    .build()
                preview.surfaceProvider = cameraView!!.getSurfaceProvider()

                val imageAnalysis = ImageAnalysis.Builder()
                    .build()
                imageAnalysis.setAnalyzer(cameraExecutor!!, barcodeImageAnalyzer)

                val cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA

                try {
                    provider.unbindAll()
                    provider.bindToLifecycle(this, cameraSelector, imageAnalysis, preview)
                } catch (e: Exception) {
                    utils.error("provider.bindToLifecycle; " + e.message)
                }
            } catch (e: Exception) {
                utils.error("cameraProvider.addListener; " + e.message)
            }
        }, ContextCompat.getMainExecutor(this))
    }

    private fun startCamera() {
        dataBaseItem = null
        textValue!!.text = ""
        textDescription!!.text = ""
        try {
            textValue!!.setText(R.string.scanning)
            setupCamera()
        } catch (_: SecurityException) {
            textValue!!.setText(R.string.error_no_permission)
        } catch (e: Exception) {
            textValue!!.text = e.toString()
        }
    }

    private fun stopCamera() {
        try {
            cameraProvider!!.get().unbindAll()
        } catch (e: Exception) {
            utils.debug("Unbind camera provider; " + e.message)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) onBackPressed()
        return super.onOptionsItemSelected(item)
    }

    override fun onDataLoaded(dataItems: java.util.ArrayList<DataBaseItem?>?) {
        progressBar!!.visibility = View.GONE
        val items = dataItems ?: ArrayList()
        if (items.isNotEmpty()) {
            dataBaseItem = items[0]
            textDescription!!.text = dataBaseItem!!.getString("description")
            textDescription!!.setTextColor(ContextCompat.getColor(this, R.color.colorPrimaryDark))
            confirmAddItem()
        } else {
            dataBaseItem = null
            textDescription!!.setText(R.string.warn_no_barcode)
            textDescription!!.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
        }
    }

    override fun onDataLoaderError(error: String?) {
        progressBar!!.visibility = View.GONE
        Toast.makeText(this, R.string.toast_error, Toast.LENGTH_SHORT).show()
    }

    private fun onBarcodeReceived(barcodeValue: String?) {
        stopCamera()

        val toneGenerator = ToneGenerator(AudioManager.STREAM_NOTIFICATION, 100)
        toneGenerator.startTone(ToneGenerator.TONE_PROP_BEEP)

        textValue!!.text = barcodeValue
        progressBar!!.visibility = View.VISIBLE

        val barcodeParameters = DataBaseItem()
        barcodeParameters.put("value", barcodeValue)
        barcodeParameters.put("guid", documentGUID)

        val dataLoader = DataLoader(this)
        dataLoader.getItemWithBarcode(barcodeParameters)
    }

    private fun confirmAddItem() {
        val intent = getIntent()
        if (dataBaseItem != null) {
            intent.putExtra("cacheKey", Cache.Companion.getInstance().put(dataBaseItem))
        } else {
            intent.putExtra("cacheKey", "")
        }
        setResult(RESULT_OK, intent)
        finish()
    }
}